src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3033,8 +3047,88 @@ void initializeLoop()
             {
                 if (isObjectLoop)
                 {
-                    // PROTOTYPE(dataflow): Handle loops over object
-                    throw new NotImplementedException();
+                    // For i as Object = 3 To 6 step 2
+                    //    body
+                    // Next
+                    //
+                    // becomes ==>
+                    //
+                    // {
+                    //   Dim loopObj        ' mysterious object that holds the loop state
+                    //
+                    //   ' helper does internal initialization and tells if we need to do any iterations
+                    //   if Not ObjectFlowControl.ForLoopControl.ForLoopInitObj(ctrl, init, limit, step, ref loopObj, ref ctrl) 
+                    //                               goto exit:
+                    //   start:
+                    //       body
+                    //
+                    //   continue:
+                    //       ' helper updates loop state and tells if we need to do another iteration.
+                    //       if ObjectFlowControl.ForLoopControl.ForNextCheckObj(ctrl, loopObj, ref ctrl) 
+                    //                               GoTo start
+                    // }
+                    // exit:
+
+                    IOperation condition;
```

Comment:
Can the code in this and checkLoopCondition() be refactored together? It seems like the only big differences is how many things are put onto the eval stack, which could be controlled by a flag?

Reply:
> Can the code in this and checkLoopCondition() be refactored together? It seems like the only big differences is how many things are put onto the eval stack, which could be controlled by a flag?

I'll se what I can do (in a follow-up PR)

---
In reply to: [188128275](https://github.com/dotnet/roslyn/pull/26803#discussion_r188128275) [](ancestors = 188128275)